### ISSUE TITLE
prange induces race conditions, parallelization slows down code dramatically with range

### ISSUE THREAD

**Issue by jrt54** [numba/numba#3522]
The toy problem here is that I'm taking an input matrix and outputting a matrix where every row is a sum of all rows in the input matrix. I implemented a "stacking" function that fills the output row of a matrix and iterated through all rows of the matrix, calling this stacking function to fill each row. 

I've tried several different methods for implementing the stacking function to test numpy's performance/accuracy. 

1.) Using np.sum
2.) Using a simple for loop with "range" and iteratively adding each row from the input to the output row
3.) Using prange with parallelization (parallel=True) and iteratively adding each row from input to the output row

I noticed that with parallelization turned off, my performance was always better across the board. The scale of performance issues is quite large in particular when I run with parallelization and range (see function: stack_parallel_range). I ran this twice in a row to see if the issue was execution or compilation, but it took 35-100x longer than any other implementation and the second run of this function was roughly the same in terms of runtime. 

Another issue is that when I ran with prange and parallelization=True, I encountered a race condition (see function: stack_parallel_prange). This was not the case for any other implementation. I would expect based off of the documentation 1.10.2 (https://numba.pydata.org/numba-doc/latest/user/parallel.html) that the use of prange with a reduction operation (+=) would not induce a race condition. Have I misread this? Does this statement not hold for reduction operations summing into arrays?

Here's a table of runtimes/errors/descriptions for the 5 experiments I ran. 

Num | Description | Implementation | Run Times | Validation Error
------------ | ------------- | ------------- | ------------- | -------------
1 | explicit range loop and no parallelization | stack | .420 secs | 0
2 | explicit range loop and parallelization | stack_parallel_range | 35.700 secs, then 35.820 secs | 0
3 | explicit prange loop and parallelization | stack_parallel_prange | .809 secs | 809426.121919
4 | np.sum and parallelization | stack_parallel_sum | 1.372 secs | 0
5 | np.sum and no parallelization | stack_sum | .260 secs | 0

So the only implementation with a race condition was the one using prange and parallelization (3). Performance issues were only particularly noticeable, at least for this operation, with a range loop and parallelization (2). However, in all cases I achieved better performance without enabling parallelization. 

Below is the script I'm using for measurements/analysis. 

```python

import numba
from numba import prange
import numpy as np
import time

@numba.njit(parallel=False)
def stack_sum(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]

    outsamples[trace_idx, :] = np.sum(insamples, axis=0)

#for some reason running with parallel flag is slower
@numba.njit(parallel=True)
def stack_parallel_sum(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]

    outsamples[trace_idx, :] = np.sum(insamples, axis=0)

#turning on parallelism with prange creates race condition
@numba.njit(parallel=True)
def stack_parallel_prange(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in prange(ntrc):
        outsamples[trace_idx, :] += insamples[i]


#this has no race conditions but is very slow
@numba.njit(parallel=True)
def stack_parallel_range(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in range(ntrc): ##using range works, but not prange, if parallelization is on this is slow 
        outsamples[trace_idx, :] += insamples[i]

#this is fast and no race conditions, but parallelization is disabled
@numba.njit(parallel=False)
def stack(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in range(ntrc): 
        outsamples[trace_idx, :] += insamples[i]




if __name__ == "__main__":


    #I'm making a 2d array of size ntrc x 1000
    #I'm trying to output an array where each row is a stacking of the input array (each row is a sum of all input rows) 
    ntrc=213
    x = np.random.rand( ntrc, 1000)
    y = np.zeros([ntrc, 1000])
    y2 = np.zeros([ntrc, 1000])
    
    
    start_time=time.time()
    
    for idx in range( x.shape[0]):
        stack(x, y, idx)
        stack(x, y2,idx)
    print("When running numba with parallelization disabled, then the stack function is very fast, the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization disabled, then the stack function does not encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )

    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_prange(x, y, idx)
        stack_parallel_prange(x, y2,idx)
    
    print("When running numba with parallelization and prange, then the stack function is very fast, the time it takes is {}".format(time.time()-start_time )  )
    print("When running numba with parallelization and prange, then the stack function does encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    
    #run again to see if cached function is faster after it has compiled
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_range(x, y, idx)
        stack_parallel_range(x, y2,idx)
    
    print("When running numba with parallelization and range for a second time, then the stack function is very slow, the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization and range for a second time, then the stack function is very slow but does not encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    

    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_range(x, y, idx)
        stack_parallel_range(x, y2,idx)
    
    print("When running numba with parallelization and range, then the stack function is very slow, the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization and range, then the stack function is very slow but does not encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )    

    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_sum(x, y, idx)
        stack_parallel_sum(x, y2,idx)
    
    print("When running numba with parallelization and numpy sum, then the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization and numpy sum, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    
    
    
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_sum(x, y, idx)
        stack_sum(x, y2,idx)
    
    print("When running numba with parallelization disabled but using numpy sum, then the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization disabled but using numpy sum, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
```

**Comment by stuartarchibald** [comment]
Thanks for the report. Before investigating the performance timings further I think a couple of things need changing to make the sample times collected comparable. Also, I think the race condition may well be a bug in parallel array reduction, ping @DrTodd13 .

1. Recall that Numba is a JIT compiler, this means the first time an `@jit` decorated function is executed for a given input type combination Numba has to compile the function to native code before running it. In the case of `@jit` this can take some time, in the case of `@jit(parallel=True)` this can take even more time. As the function being executed runs in a time frame similar to that of compilation time it's very hard to obtain an accurate measurement for run time (the compilation cost is never amortized). [This is probably worth a read for additional guidance](http://numba.pydata.org/numba-doc/latest/user/5minguide.html#how-to-measure-the-performance-of-numba).
2. I think functions like this:
    ```python
    @numba.njit(parallel=False)
    def stack_sum(insamples, outsamples, trace_idx):
        outsamples[trace_idx,:]=0.0
        ntrc=outsamples.shape[0]
    
        outsamples[trace_idx, :] = np.sum(insamples, axis=0)
    ```
    are doing unnecessary work in that `outsamples[trace_idx,:]` is first set to zero, and then overwritten by the result of the summation over `insamples`? In the cases where a manual reduction using `+=` takes place this is necessary, but the `np.sum` use has this implied.

**Comment by thoth291** [comment]
With all the respect to the correctness of the comments, @stuartarchibald, I would say that overhead for 35 seconds seems to be not related to compilation time and unnecessary work in stack_sum doesn't stop it from being the fastest among all the implementations. But yes, I would say that compilation time should be avoided during proper benchmarks.

**Comment by jrt54** [comment]
With respect to 1.) I agree that I should have run every function twice and only recorded the second run. I did this only in the case of the 35 second run in the same way suggested in that link. That run was the biggest surprise to me. 

If the issue causing the 35 second runtime for experiment 2 was due to the JIT compiler, then I would expect it to be much faster on the second run. But it was 35.7 then 35.8 seconds, so if the JIT compiler is causing then the compilation is not being cached.

**Comment by stuartarchibald** [comment]
@thoth291 @jrt54 Thanks for your responses. The advice offered, to be aware of the JIT when recording timings, is just standard advice offered with view of raising user awareness of this issue. This advice is applicable invariant of any performance bugs/issues. Essentially, this needs fixing in the sample code so it produces comparable results, but it doesn't necessarily impact the result which has a performance bug.

**Comment by stuartarchibald** [comment]
As to this function:
```python
#turning on parallelism with prange creates race condition
@numba.njit(parallel=True)
def stack_parallel_prange(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in prange(ntrc):
        outsamples[trace_idx, :] += insamples[i]
```
I think that this is probably what is happening, the `prange` loop runs  `0..ntrc` in the induction variable `i`, each thread will take a value of `i` and all of them will try and read from and write into `outsamples[trace_idx, :]`. I think at present the underlying technology cannot spot this pattern and rewrite it as a parallel reduction. To quote [the documentation](http://numba.pydata.org/numba-doc/dev/user/parallel.html#explicit-parallel-loops):
> A reduction is inferred automatically if a variable is updated by a binary function/operator using its previous value in the loop body. The initial value of the reduction is inferred automatically for += and *= operators. For other functions/operators, the reduction variable should hold the identity value right before entering the prange loop. Reductions in this manner are supported for scalars and for arrays of arbitrary dimensions.

The important part being `Reductions in this manner are supported for scalars and for arrays of arbitrary dimensions.`, which doesn't mention slices or indexing works with array reduction. @DrTodd13 is this still the case?

This seems to do the intended without a race condition:
```python
@numba.njit(parallel=True)
def stack_parallel_prange(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    write_to = np.zeros((outsamples.shape[1],))
    for i in prange(ntrc):
        write_to += insamples[i]
    outsamples[trace_idx, :] = write_to
```

As to the apparent performance bug in `stack_parallel_range`... the new parallel diagnostics tool which will be available from the upcoming 0.41.0 release, shows this:
```
================================================================================
 Parallel Accelerator Optimizing:  Function stack_parallel_range, issue3522_b.py
 (38)  
================================================================================


Parallel loop listing for  Function stack_parallel_range, issue3522_b.py (38) 
---------------------------------------------------------------|loop #ID
@numba.njit(parallel=True)                                     | 
def stack_parallel_range(insamples, outsamples, trace_idx):    | 
    outsamples[trace_idx,:]=0.0--------------------------------| #0
    ntrc=outsamples.shape[0]                                   | 
    for i in range(ntrc):                                      | 
        outsamples[trace_idx, :] += insamples[i]---------------| #1
```
The `outsamples[trace_idx,:]=0.0` operation is parallelized (parallel loop `#0`), as is the **body** of the `range` loop (parallel loop `#1`). I would expect the cause of the apparent slowness of this function to be down to repeatedly running a small amount of parallel work in the `range` loop. Every iteration of that loop will involve launching multiple parallel kernels that have very little work to do, I'd guess that this launch overhead is probably the reason for the slow down.

**Comment by jrt54** [comment]
Thank you for the feedback. I've updpated the code now and posted it at the bottom. It should now only be comparing runtimes after compilation has been done. I'm also adding a control here which is the simple range loop without numba for whatever that's worth. On my end, your fix did indeed resolve the race condition. Here's the table of runtimes/results:

Num | Description | Implementation | Run Times | Validation Error
------------ | ------------- | ------------- | ------------- | -------------
1 | explicit range loop and no parallelization | stack | .068 secs | 0
2 | explicit range loop and parallelization | stack_parallel_range |35.635 | 0
3 | explicit prange loop and parallelization | stack_parallel_prange | .352 secs | 796169.908695
4 | fixed explicit prange loop and parallelization | stack_parallel_prange_fix | .749 secs | 0
5 | np.sum and parallelization | stack_parallel_sum | .201 | 0
6 | np.sum and no parallelization | stack_sum | .029 | 0
7 | range and no numba at all | stack_control | .116 | 0


The issue I see still is that parallelization has added nothing in terms of performance anywhere, even if I can fix the race conditions. Understandably, this is a simple operation anyway and maybe I stand little to gain anyway from parallelization, but is there a better way to design these functions so that I do gain something from the parallelization? Or do I only stand something to gain if I'm doing something much more arithmteically-intense within a prange loop? 

And the script: 
```python
import numba
from numba import prange
import numpy as np
import time

@numba.njit(parallel=False)
def stack_sum(insamples, outsamples, trace_idx):
    outsamples[trace_idx, :] = np.sum(insamples, axis=0)

#for some reason running with parallel flag is slower
@numba.njit(parallel=True)
def stack_parallel_sum(insamples, outsamples, trace_idx):
    outsamples[trace_idx, :] = np.sum(insamples, axis=0)

#this works but is very slow
@numba.njit(parallel=True)
def stack_parallel_range(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in range(ntrc): ##using range works, but not prange, if parallelization is on this is slow but it's fast if parallel=False
        outsamples[trace_idx, :] += insamples[i]


#turning on parallelism with prange creates race condition
@numba.njit(parallel=True)
def stack_parallel_prange(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in numba.prange(ntrc):
        outsamples[trace_idx, :] += insamples[i]

#fixing prange
@numba.njit(parallel=True)
def stack_parallel_prange_fix(insamples, outsamples, trace_idx):
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    write_to = np.zeros((outsamples.shape[1],))
    for i in prange(ntrc):
        write_to += insamples[i]
    outsamples[trace_idx, :] = write_to

#this is fast and no race conditions, but parallelization is disabled
@numba.njit(parallel=False)
def stack(insamples, outsamples, trace_idx):
    #outsamples[rec,trace_idx,:]=0.0
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in range(ntrc): ##using range works, but not prange, if parallelization is on this is slow but it's fast if parallel=False
        outsamples[trace_idx, :] += insamples[i]

#no numba
def stack_control(insamples, outsamples, trace_idx):
    #outsamples[rec,trace_idx,:]=0.0
    outsamples[trace_idx,:]=0.0
    ntrc=outsamples.shape[0]
    for i in range(ntrc): ##using range works, but not prange, if parallelization is on this is slow but it's fast if parallel=False
        outsamples[trace_idx, :] += insamples[i]


# putting this at the end also works and probably auto-parallelizes
#    outsamples[trace_idx, :] = np.sum(insamples, axis=0)
if __name__ == "__main__":


    #I'm making a 2d array of size ntrc x 1000
    #I'm trying to output an array where each row is a stacking of the input array (each row is a sum of all input rows) 
    ntrc=213
    x = np.random.rand( ntrc, 1000)
    y = np.zeros([ntrc, 1000])
    y2 = np.zeros([ntrc, 1000])
    
    #pre-run so as not to effect runtime 
    stack(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    
    for idx in range( x.shape[0]):
        stack(x, y, idx)
        stack(x, y2,idx)
    print("When running numba with parallelization disabled, then the stack function is very fast, the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization disabled, then the stack function does not encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )

    stack_parallel_prange(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_prange(x, y, idx)
        stack_parallel_prange(x, y2,idx)
    
    print("When running numba with parallelization and prange, then the stack function is very fast, the time it takes is {}".format(time.time()-start_time )  )
    print("When running numba with parallelization and prange, then the stack function does encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    
    stack_parallel_prange_fix(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_prange_fix(x, y, idx)
        stack_parallel_prange_fix(x, y2,idx)
    
    print("When running numba with parallelization and prange_fix, then the stack function is very fast, the time it takes is {}".format(time.time()-start_time )  )
    print("When running numba with parallelization and prange_fix, then the stack function does not encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    
    
    stack_parallel_range(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_range(x, y, idx)
        stack_parallel_range(x, y2,idx)
    
    print("When running numba with parallelization and range, then the stack function is very slow, the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization and range, then the stack function is very slow but does not encounter a race condition, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    
    stack_parallel_sum(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_parallel_sum(x, y, idx)
        stack_parallel_sum(x, y2,idx)
    
    print("When running numba with parallelization and numpy sum, then the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization and numpy sum, the difference between runs is {}".format(np.sum(abs(y-y2)))  )
    
    
    
    stack_sum(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_sum(x, y, idx)
        stack_sum(x, y2,idx)
    
    print("When running numba with parallelization disabled but using numpy sum, then the time it takes is {}".format(time.time()-start_time ))
    print("When running numba with parallelization disabled but using numpy sum, the difference between runs is {}".format(np.sum(abs(y-y2)))  )


    stack_control(x, y, 0)
    y = np.zeros([ntrc, 1000])
    start_time=time.time()
    for idx in range( x.shape[0]):
        stack_control(x, y, idx)
        stack_control(x, y2,idx)

    print("When running without numba,  then the time it takes is {}".format(time.time()-start_time ))
    print("When running without numba, the difference between runs is {}".format(np.sum(abs(y-y2)))  )

```

**Comment by stuartarchibald** [comment]
Closing this as I think we got to the bottom of all the issues. If this is not the case please open a new ticket for a distinct issue. Thanks.
